Adding Controller user's guide (#77)
nephio-project/docs: content/en/docs/guides/user-guides/controllers.md

---
title: Nephio controller user's guide
description: >

weight: 1
---

## Reconcilers

[Reconcilers](https://kubebyexample.com/learning-paths/operator-framework/operator-sdk-go/controller-reconcile-function)
are used by Kubernetes to enforce the desired state of a Custom Resource (CR).

## The Nephio Reconciler Interface

> The Nephio Controller supports multiple reconcilers and allows pluggable reconcilers. To plug into the Nephio controller, a reconciler implements the

Reviewer: ```suggestion
The Nephio Controller supports multiple reconcilers and allows pluggable reconcilers. To plug into the Nephio controller, a reconciler implements the
```

> The Nephio Controller supports multiple reconcilers and allows pluggable reconcilers. To plug into the Nephio controller, a reconciler implements the

Reviewer: ```suggestion
The Nephio Controller deploys multiple reconcilers and also supports pluggable reconcilers. To plug into the Nephio controller, a reconciler must implement the ...
```

> The Nephio Controller deploys multiple reconcilers and also supports pluggable reconcilers. To plug into the Nephio controller, a reconciler must implement the ...

Reviewer 2: is the trailing '..." a typo?

Author: Yes, fixed in [8f19d0a](https://github.com/nephio-project/docs/pull/77/commits/8f19d0ac88ceca08c3456b92e0fb5aa7f22ad25d)

[nephio reconciler
interface](https://github.com/nephio-project/nephio/tree/main/controllers/pkg/reconcilers/reconciler-interface).
Reconcilers register with the Nephio reconciler interface when they start up.

> The reconcilers below are currently supported in the nephio controller:

Reviewer: ```suggestion
The reconcilers below are currently deployed by default in the nephio controller:
```

```bash
./controllers/pkg/reconcilers/token/reconciler.go
./controllers/pkg/reconcilers/repository/reconciler.go
./controllers/pkg/reconcilers/bootstrap-packages/reconciler.go
./controllers/pkg/reconcilers/vlan-specializer/reconciler.go
./controllers/pkg/reconcilers/generic-specializer/reconciler.go
./controllers/pkg/reconcilers/network/reconciler.go
./controllers/pkg/reconcilers/ipam-specializer/reconciler.go
./controllers/pkg/reconcilers/bootstrap-secret/reconciler.go
./controllers/pkg/reconcilers/approval/reconciler.go
```

## Enabling Reconcilers

To enable a particular reconciler, you pass an environment variable to the
Nephio Controller at startup. The environment variable is of the form

> `ENABLE-<RECONCILER>` where `<RECONCILER>` is the name of the reconciler to

Reviewer: ```suggestion
`ENABLE_<RECONCILER>`  ...
```

Author: Sorry, but this suggestion is not clear. With the suggestion the sentence would look like this: 
"The environment variable is of the form `ENABLE_<RECONCILER>` ... be enabled in upper case."

Reviewer: Minor fix. I think the env var should use underscore instead of hyphen.
"of the form `ENABLE_<RECONCILER>`

Author: Fixed in [9b070b4](https://github.com/nephio-project/docs/pull/77/commits/9b070b417343ded61102818f27e111e74f3ba646)

Containers:
  controller:
    Container ID:  containerd://37d3eff53c1944a659e5a7ab913173db42f34b44347072e6c9b51e5671f35ea2
    Environment:
      POD_NAMESPACE:              nephio-system (v1:metadata.namespace)
      POD_IP:                      (v1:status.podIP)
      POD_NAME:                   nephio-controller-6565fd695d-44kld (v1:metadata.name)
      NODE_NAME:                   (v1:spec.nodeName)
      NODE_IP:                     (v1:status.hostIP)
      GIT_URL:                    http://172.18.0.200:3000
      GIT_NAMESPACE:              gitea
      ENABLE_APPROVAL:            true
      ENABLE_REPOSITORIES:        true
      ENABLE_BOOTSTRAPSECRETS:    true
      ENABLE_BOOTSTRAPPACKAGES:   true
      ENABLE_GENERICSPECIALIZER:  true
      ENABLE_NETWORKS:            true
      CLIENT_PROXY_ADDRESS:       resource-backend-controller-grpc-svc.backend-system.svc.cluster.local:9999
```

</details>

To check that the reconcilers are actually deployed, you can examine the logs
from the Nephio Controller. The log rolls over so you may need to redeploy the
Nephio Controller to see what reconcilers are being deployed.

```bash

>  $ kubectl rollout restart deployment nephio-controller -n nephio-system

Reviewer: unindent

Author: Fixed in [afd20ba](https://github.com/nephio-project/docs/pull/77/commits/afd20ba7ec89e31a504f7fc431c0a52bc9dfa89a)

> $ k logs -n nephio-system nephio-controller-59487989bf-md845 --all-containers | grep enable

Reviewer: Use kubeclt to keep it consistent and maybe push the greped response to a new line

Author: Fixed in [afd20ba](https://github.com/nephio-project/docs/pull/77/commits/afd20ba7ec89e31a504f7fc431c0a52bc9dfa89a)

> grep enable 2023-06-27T11:37:58.646Z	INFO	setup	enabled reconcilers	{"reconcilers": "repositories,approval,bootstrappackages,bootstrapsecrets,genericspecializer,networks"}

Reviewer 2: ```suggestion
grep enable
2023-06-27T11:37:58.646Z	INFO	setup	enabled reconcilers	{"reconcilers": "repositories,approval,bootstrappackages,bootstrapsecrets,genericspecializer,networks"}
```

Reconcilers register with the Nephio reconciler interface when they start up.

The reconcilers below are currently deployed by default in the nephio controller:

```bash
./controllers/pkg/reconcilers/token/reconciler.go
./controllers/pkg/reconcilers/repository/reconciler.go
./controllers/pkg/reconcilers/bootstrap-packages/reconciler.go
./controllers/pkg/reconcilers/vlan-specializer/reconciler.go
./controllers/pkg/reconcilers/generic-specializer/reconciler.go
./controllers/pkg/reconcilers/network/reconciler.go
./controllers/pkg/reconcilers/ipam-specializer/reconciler.go
./controllers/pkg/reconcilers/bootstrap-secret/reconciler.go
./controllers/pkg/reconcilers/approval/reconciler.go
```

## Enabling Reconcilers

To enable a particular reconciler, you pass an environment variable to the
Nephio Controller at startup. The environment variable is of the form
`ENABLE_<RECONCILER>` where `<RECONCILER>` is the name of the reconciler to
be enabled in upper case. Therefore, to enable the `bootstrap-packages` reconciler,
pass the `ENABLE_BOOTSTRAPPACKAGES` to the nephio controller. Reconcilers are

> disabled by default.

Reviewer: Should this read "enabled by default" ?

> <details>

Reviewer: Do you want to remove the drop down here?

Author: Not yet. Let's agree in https://github.com/nephio-project/nephio/issues/495 where we remove them and we will create separate pr-s based on that.
